network/modules/tcp/server_socket.py

```diff
@@ -0,0 +1,114 @@
+"""
+Wrapper for TCP server socket operations.
+"""
+
+import socket
+
+from .socket_wrapper import TcpSocket
+
+
+class TcpServerSocket(TcpSocket):
+    """
+    Wrapper for TCP server socket operations.
+    """
+
+    __create_key = object()
+
+    def __init__(self, class_private_create_key: object, socket_instance: socket.socket) -> None:
+        """
+        Private constructor, use create() method.
+        """
+
+        assert class_private_create_key is TcpServerSocket.__create_key, "Use create() method"
+
+        super().__init__(socket_instance=socket_instance)
+
+    @classmethod
+    def create(
+        cls,
+        instance: socket.socket = None,
+        host: str = "",
+        port: int = 5000,
+        connection_timeout: float = 60.0,
+    ) -> "tuple[bool, TcpServerSocket | None]":
+        """
+        Establishes socket connection through provided host and port.
+            Note: Although in practice a TCP 'server' simply connects 2 clients together,
+            this newly created client is called 'server' for simplicity and differentiation.
+
+        Parameters
+        ----------
+        instance: socket.socket (default None)
+            For initializing Socket with an existing socket object.
+        host: str (default "")
+            Empty string is interpreted as '0.0.0.0' (IPv4) or '::' (IPv6), which is all addresses.
+            Could also use socket.gethostname(). (needed to enable other machines to connect)
+        port: int (default 5000)
+            The host combined with the port will form an address (e.g. localhost:5000)
+        connection_timeout: float (default 10.0)
+            Timeout for operations such as recieve
+
+        Returns
+        -------
+        tuple[bool, TcpServerSocket | None]
+            The first parameter represents if the socket creation is successful.
+            - If it is not successful, the second parameter will be None.
+            - If it is successful, the second parameter will be the created
+              TcpServerSocket object.
+        """
+
+        # Reassign instance before check or Pylance will complain
+        socket_instance = instance
+        if socket_instance is not None:
+            return True, TcpServerSocket(cls.__create_key, socket_instance)
+
+        if socket.has_dualstack_ipv6():
```

Comment:
seems like this depends on the library - if we use a specific version would we be able to know if it can accept both IPv4 and IPv6? or is this smth that has to do with the hardware running the server?

Reply:
This is hardware/OS related.